### ISSUE TITLE
[Bug] index_put_ overflows 32-bit element offsets for large strided tensors

### ISSUE THREAD

**Issue by ZenithJoyH** [flagos-ai/FlagGems#6411]
## Summary

The generic `flag_gems.index_put_` Triton kernel silently writes to the wrong location when the computed element offset exceeds signed int32. This is reproducible on a T-Head PPU with a production Qwen3.8 GDN state layout. Native ATen handles the identical tensor, stride, int32 index, and value correctly.

This has a model-level accuracy impact: in a full GPQA Diamond run (198 samples, concurrency 64), enabling only the FlagGems `index_put_` ATen replacement reduced strict accuracy from 173/198 (87.37%, native ATen baseline) to 119/198 (60.10%).

## Environment

- FlagGems revision: `2aacb2ed7b116baee0acc147cdc770b4176a6cdd` (`support-qwen3.8-flash`)
- PyTorch device API: CUDA-compatible
- Accelerator: T-Head PPU-ZW810E
- Input dtype: `torch.float32`
- Index dtype: `torch.int32`
- `accumulate=False`

## Production call captured from vLLM

Qwen3.8 performs a GDN state update with:

- input shape: `[80884, 6, 128, 128]`
- input stride: `[102400, 16384, 128, 1]`
- logical input size: about 31.8 GB
- values shape: `[6, 128, 128]`
- operation: `ssm_state[prefill_state_indices] = last_recurrent_state`

The first row whose element offset exceeds signed int32 is `ceil(2^31 / 102400) = 20972`.

## Minimal reproducer

The reproducer truncates dim 0 at the first failing row, so it needs about 8.6 GiB of device memory instead of the full 33 GiB backing storage.

```python
import gc
import math
import torch
import flag_gems

ROW_SHAPE = (6, 128, 128)
STATE_STRIDE = (102400, 16384, 128, 1)
FIRST_INT32_OVERFLOW_ROW = math.ceil((2**31) / STATE_STRIDE[0])
STATE_SHAPE = (FIRST_INT32_OVERFLOW_ROW + 1, *ROW_SHAPE)

def test_qwen38_gdn_index_put_large_offset():
    assert FIRST_INT32_OVERFLOW_ROW == 20972
    state = torch.empty_strided(
        STATE_SHAPE, STATE_STRIDE, dtype=torch.float32, device="cuda"
    )
    host_index = torch.tensor(
        [FIRST_INT32_OVERFLOW_ROW], dtype=torch.int32, pin_memory=True
    )
    index = host_index.to("cuda", non_blocking=True)
    base = torch.arange(128, dtype=torch.float32, device="cuda").view(1, 1, 128)
    native_value = base.expand(ROW_SHAPE).clone()
    gems_value = native_value + 10000

    try:
        state.index_put_((index,), native_value, accumulate=False)
        torch.cuda.synchronize()
        torch.testing.assert_close(
            state[index], native_value.unsqueeze(0), rtol=0, atol=0
        )

        flag_gems.index_put_(state, (index,), gems_value, accumulate=False)
        torch.cuda.synchronize()
        torch.testing.assert_close(
            state[index], gems_value.unsqueeze(0), rtol=0, atol=0
        )
    finally:
        del state
        gc.collect()
        torch.cuda.empty_cache()
```

Run with:

```bash
CUDA_VISIBLE_DEVICES=0 pytest -vv -s test_index_put_gdn_large_offset_minimal.py
```

## Actual result

Native ATen passes. FlagGems fails deterministically after an explicit device synchronize:

```text
Mismatched elements: 98304 / 98304 (100.0%)
Greatest absolute difference: 10000.0 at index (0, 0, 0, 0)
```

Low rows `0..31` and 64 repeated low-slot updates pass. A mixed test using rows `(0, 5243, 20972, 41944, 80883)` fails only for the final three rows, which are the rows whose element offsets exceed `2^31`.

## Suspected cause

The generated kernel builds `input_offset` from expressions such as `cur_index0 * input_stride0`. With an int32 index and a stride value that itself fits int32, Triton appears to perform the multiplication with a 32-bit intermediate. The result overflows before pointer arithmetic.

All input, value, and index offset arithmetic should be promoted to `tl.int64` before multiplication/addition. The explicit synchronize in the reproducer rules out a stream-ordering race.

## Locally validated fix

Promoting the generated program offsets and loaded tensor indices to `tl.int64` before stride multiplication fixes the failure:

```python
offset0 = pid0.to(tl.int64) * BLOCK_SIZE0 + tl.arange(0, BLOCK_SIZE0).to(tl.int64)[:, None]
cur_index0 = tl.load(...).to(tl.int64)
```

Validation after this change:

- compact 2.15 GiB boundary reproducer: passed
- production-stride 8.6 GiB boundary reproducer: passed
- complete 31.8 GB GDN layout suite: 3/3 passed
- eager, non-blocking H2D, CUDA Graph, ATen dispatch, non-contiguous, and `torch.compile` coverage: 16/16 passed
- focused source regression test, `git diff --check`, and ruff checks: passed
